redaxo/src/core/lib/sql/table.php

```diff
@@ -141,56 +172,150 @@ public function removeColumn($name)
+
+        $this->primaryKey = (array) $columns;
+        $this->primaryKeyModified = true;
+
+        return $this;
+    }
+
+    public function create()
+    {
+        $parts = [];
+
+        foreach ($this->columns as $column) {
+            $parts[] = $this->getColumnDefinition($column);
+        }
+
+        if ($this->primaryKey) {
+            $parts[] = 'PRIMARY KEY '.$this->getKeyColumnsDefintion($this->primaryKey);
+        }
+
+        $query = 'CREATE TABLE '.$this->sql->escapeIdentifier($this->name)." (\n    ";
+        $query .= implode(",\n    ", $parts);
+        $query .= "\n) ENGINE=InnoDB DEFAULT CHARSET=utf8;";
+
+        $this->sql->setQuery($query);
+
+        $this->resetModified();
+    }
+
     public function alter()
     {
-        $queries = [];
-        $sql = rex_sql::factory();
-
-        $columnDefinition = function (rex_sql_column $column) use ($sql) {
-            return sprintf(
-                '%s %s %s %s',
-                $sql->escapeIdentifier($column->getName()),
-                $column->getType(),
-                $column->getDefault() ? 'DEFAULT '.$sql->escape($column->getDefault()) : '',
-                $column->isNullable() ? '' : 'NOT NULL',
-                $column->getExtra()
-            );
-        };
+        $parts = [];
+
+        if ($this->primaryKeyModified) {
+            $parts[] = 'DROP PRIMARY KEY';
+        }
 
         $columns = $this->columns;
-        foreach ($this->existing as $name) {
+        foreach ($this->columnsExisting as $name) {
             if (!isset($columns[$name])) {
-                $queries[] = 'DROP '.$sql->escapeIdentifier($name);
+                $parts[] = 'DROP '.$this->sql->escapeIdentifier($name);
                 continue;
             }
 
             $column = $columns[$name];
             if ($column->isModified()) {
-                $queries[] = 'CHANGE '.$sql->escapeIdentifier($name).' '.$columnDefinition($column);
+                $parts[] = 'CHANGE '.$this->sql->escapeIdentifier($name).' '.$this->getColumnDefinition($column);
             }
             unset($columns[$name]);
         }
         foreach ($columns as $column) {
-            $queries[] = 'ADD '.$columnDefinition($column);
+            $parts[] = 'ADD '.$this->getColumnDefinition($column);
         }
 
-        if (!$queries) {
+        if ($this->primaryKeyModified && $this->primaryKey) {
+            $parts[] = 'ADD PRIMARY KEY '.$this->getKeyColumnsDefintion($this->primaryKey);
+        }
+
+        if (!$parts) {
             return;
         }
 
-        $query = 'ALTER TABLE '.$sql->escapeIdentifier($this->name)."\n    ";
-        $query .= implode(",\n    ", $queries);
+        $query = 'ALTER TABLE '.$this->sql->escapeIdentifier($this->name)."\n    ";
+        $query .= implode(",\n    ", $parts);
         $query .= ';';
 
-        $sql->setQuery($query);
+        $this->sql->setQuery($query);
+
+        $this->resetModified();
+    }
+
+    public function ensure()
```

Comment:
ich würde die methoden weiter oben hinsortieren die man "öfter brauch", d.h. ensure() und drop() ist in meinen augen die "haupt-api". dass man explizit "alter" oder "create" benutzt ist vermutlich eher selten im vergleich